notes/smart-pointers-1.md

```diff
@@ -0,0 +1,244 @@
+}
+
+```
+
+---
+## `unique_ptr` Operations
+
+For unique pointer `x`:
+
+`x.get()` returns a pointer to the pointed-to object
+
+`*x` returns value of pointed-to object
+- Basically normal dereference
+
+`x.reset(ptr)`: run `delete` on the current owned pointer, then store a new one, `ptr`
+
+`x.release()` returns the pointer, sets wrapped pointer to `nullptr`
+- Releases responsibility for explicitly freeing back to user
+
+Example:
+```C++
+#include <memory> // unique_ptr
+#include <iostream> // cout
+#include <cstdlib> // EXIT_SUCCESS
+
+int main(){
+  // x is a unique pointer that owns an int*
+  std::unique_ptr<int> x(new int(333));
+
+  // x.get() returns pointer to pointer-to object
+  int* ptr = x.get();
+  std::cout << *x << std::endl; // prints 333
+
+  // delete current pointer, stores a new one
+  x.reset(new int(124));
+  std::cout << *x << std::endl; // prints 124
+
+  // smart pointer releases responsibility for freeing back to user
+  ptr = x.release();
+  delete ptr;
+
+  return EXIT_SUCCESS;
+}
+
+```
+
+---
+## `unique_ptr` is not copyable
+
+As its name implies, `std::unique_ptr` cannot be copied
+- Copy constructor and copy-assignment operator are disabled
+- Compiler error if you try
+
+Bugged example:
+```C++
+std::unique_ptr<int> x(new int(333));
+std::unique_ptr<int> y(x); // Compiler error - cctor disabled
+
+std::unique<int> z; // z is nullptr
+z = x; // Compiler error - copy-assignment disabled
+
+```
+
+---
+## Bug #1
+
+We've seen that you can't copy unique pointers, but you can still wrap the same raw pointer with multiple unique pointers. This causes a problem
+
+Example:
+```C++
+int main() {
+    int* p = new int(333);
+    std::unique_ptr<int> x(p);
+    std::unique_ptr<int> y(p);
+
+    // What happens here at the end of main?
+}
+```
+- Now both `x` and `y` "own" pointer `p`
+- When `y` falls out of scope, it will call `delete` on `p`, which it owns
+- But then when `x` falls out of scope, it also will call `delete` on `p`, which has already been deleted!
+- Causes a double-free!
+
+---
+## Bug #2
+
+Also remember that smart pointers are intended to wrap raw pointers to HEAP-ALLOCATED memory
+
+You should never try to `delete` memory on the stack anyways
+
+Bugged example:
+```C++
+int main(){
+  int x = 333;
+  std::unique_ptr<int> p(&x); // BUG
+}
+```
+- This will still compile
+- However, `&x` is a pointer to memory on the stack, so trying to `delete` it will cause undefined behavior
+
+
+---
+## Transferring ownership
+
+You can transfer ownership of a pointer from one `unique_ptr` to another using `release()` and `reset()` in combination
+
+```C++
+int main() {
+    std::unique_ptr<int> x(new int(333));
+
+    std::unique_ptr<int> y(x.release());
+    // y now owns x's old pointer
+    x.reset(new int(124));
+
+    std::cout << *x << std::endl; // prints 124
+    std::cout << *y << std::endl; // prints 333
+
+}
+```
+
+You can also use move semantics (will cover later)
```

Comment:
Perhaps don't mention the code for now and put it to the other section? I do have things like this in mine as well, but I find it best just to mention it in passing.